Add layer button

elpaso opened the pull request:
## Description

The unified add layers button and dialog as discussed in Essen (see: https://docs.google.com/document/d/1aMX9jOfl10q8oETRzOHSHnRpHCc7UJ1dDnCCq3VyTAY/edit )
![unibutton-3](https://cloud.githubusercontent.com/assets/142164/26438850/af12a47c-4125-11e7-9d86-246397a27b2e.gif)

## TODO

- [x] Raster layers
- [x] Virtual layers
- [x] Delimited text layers
- [x] GUI Styles
- [x] Dialogs sizing
- [x] Left icon bar collapsible
- [x] Delimited text dialog refactoring (this is really too wide and a bit confusing)
- [x] Vector layers dialog sizing (fixed width: does not adapt to widget container)
- [x] Persist dialog status (active tab etc.)
- [x] Settings option for modeless Dialog
- [x] Methods and key bindings to open the dialog on a particular selection page
~~Dockable? See questions below~~ <<< nobody wants it


## Open questions

There are still some open questions where I'd like some feedback:

~~1 Should the dialog be also available as a dockable panel? If yes, what's the best UX pattern to let the users choose what behavior they want (a global setting, a GUI checkbox inside the dialog, a push button ...) ?~~ <<< nobody wants it
2. I would like to keep the current layers toolbar available (maybe disabled by default) power users may need it, in this case the unified button should probably stay in its own toolbar to allow for customization
3. For the same reason I'd like to keep the menu entries and their shortcuts.
4. for both 2 and 3, we can either keep the current behavior (they open the individual provider dialog) or turn it into opening the unified dialog wtith the correct "tab" open


Looking forward to hear your feedback!

elpaso committed:
[addlayerbutton] Initial implementation

elpaso committed:
[addlayerbutton] Add embedded flag for widget-like usage

elpaso committed:
[addlayerbutton] Added WFS, WMS and vector layers

elpaso committed:
[addlayerbutton] Added embeddedMode to all providers

Also switched managerMode and embeddedMode arguments
for a more logical arg passing from the factory function.

elpaso committed:
[addlayerbutton] Add embedded option to spatialite dlg

elpaso committed:
[addlayerbutton] Pass canvas to the manager (needed by AMS)

elpaso committed:
[addlayerbutton] Embedded dialog support

elpaso committed:
[addlayerbutton] Added more providers and get canvas from ctor

elpaso committed:
[addlayerbutton] New signals for new interfaces

elpaso committed:
[addlayerbutton] More embeddable dialogs

elpaso committed:
[addlayerbutton] Show Add button in embedded spatialite dialog

elpaso committed:
[addlayerbutton] Removed merge markers

jonnyforestGIS commented:
Hi @elpaso ,
great :+1: 

My opinion:

In question 1, perhaps put as an option checkbox in settings> tab data sources like the open attribute table in a dock windows?

I agree with you in question 2 and 3, for example sometimes i need to customize the GUI for non-gis users.

For question 4, if the setting checkbox (option checkbox that i refer in question 1) is active will show the unified layers if not keep the individual provider dialog.

Cheers
João

luipir commented:
great @elpaso 
would make sense to simplify the GUI having only the browser and move the left column (the list of possible providers) to a (+) button shown when mouse over the provider three item?

elpaso committed:
[addlayerbutton] Use QgsOptionsDialogBase and all its nice features and styles

timlinux commented:
@elpaso thanks so much for making this! Its looking really good!

> There are still some open questions where I'd like some feedback:

> Should the dialog be also available as a dockable panel? If yes, what's the best UX pattern to let the users choose what behavior they want (a global setting, a GUI checkbox inside the dialog, a push button ...) ?

I'd be -1 for that - the dialog isn't going to fit well in dock and we already have the browser for quickly dropping a good proportion of data sources into the canvas so it seems redundant to make this a dock widget too...

> I would like to keep the current layers toolbar available (maybe disabled by default) power users may need it, in this case the unified button should probably stay in its own toolbar to allow for customization

+1 from me to make it available still as an option (default hidden)

> For the same reason I'd like to keep the menu entries and their shortcuts.
for both 2 and 3, we can either keep the current behavior (they open the individual provider dialog) or turn it into opening the unified dialog wtith the correct "tab" open

+1 from me

Thanks!

nyalldawson commented on src/app/qgsdatasourcemanagerdialog.h:
```diff
@@ -0,0 +1,74 @@
+  class QgsDataSourceManagerDialog;
+}
+
+class QgsDataSourceManagerDialog : public QgsOptionsDialogBase
```
Can we move this to GUI? I think there's a strong case for allowing plugins to show a generic "select ANY available vector/raster" layer dialog

elpaso replied:
@nyalldawson  I'm trying to move it to `gui` but the dialog needs the browser panel which is in `app`, what would you recommend to do?

nyalldawson replied:
I'd suggest moving that widget across to gui too - I think it'd be handy to have accessible to plugins too.

elpaso replied:
I agree. But this is out of scope for this PR. 
The road to a refactoring of the provider dialogs is paved with obstacles. 
BTW I'll see if I can get some funds for this in a future iteration.

wonder-sk replied:
I would suggest that we add a new slot updateProjectHome() to QgsBrowserDockWidget. Its body would read something like this: if (mModel) mModel->updateProjectHome(). Then, the connect() would be moved from QgsBrowserDockWidget::showEvent() to QgisApp constructor where the browser dock widget gets created (there would be two connects, one for each browser dock). These few changes should remove the dependency on QgisApp...

elpaso replied:
@wonder-sk sounds good! Thanks

elpaso replied:
@nyalldawson @wonder-sk  now that I've moved the browser to `gui` I'll probably be able to work on moving the new layers dialog to `gui` in the next weeks. One problem I see is that there is a dependency on `QgsCanvas` for the ArcGIS provider that needs to be solved, I asked @manisandro about that but I've got no answer. 
https://github.com/qgis/QGIS/blob/master/src/app/qgsdatasourcemanagerdialog.cpp#L112
I'm not familiar with that provider, do you have any hint?

manisandro replied:
@elpaso Ups must have missed the notification, I'll try to take a look ASAP.

nyalldawson replied:
@manisandro While we're talking about the afs provider, any chance you could get sourcepole to look into removing https://github.com/qgis/QGIS/blob/master/src/providers/arcgisrest/qgsafsfeatureiterator.h#L33 ? It's presence makes the provider very crashy, since iterator lifetime is not tied to the lifetime of a provider and it's essential that iterators can continue to work correctly when the provider is removed.

manisandro replied:
@nyalldawson I'll take a look!

manisandro replied:
@nyalldawson Looks like the solution is to introduce a shared struct analogous to `QgsWFSSharedData`. I'll work on it.

nyalldawson replied:
Sounds perfect - thanks!

elpaso replied:
@nyalldawson @wonder-sk I'm looking at moving this class to GUI, but I've found more dependencies, like `qgsnewogrconnection.h` and all its companions under `app/ogr`, do you think that worths the effort moving all ogr deps to GUI too? I start to doubt about that.

elpaso replied:
@nyalldawson moved to GUI in https://github.com/qgis/QGIS/pull/4769

nyalldawson commented on src/core/qgsproviderregistry.h:
```diff
@@ -98,7 +98,7 @@ class CORE_EXPORT QgsProviderRegistry
      */
     QWidget *createSelectionWidget( const QString &providerKey,
-                                    QWidget *parent = nullptr, Qt::WindowFlags fl = Qt::WindowFlags() );
+                                    QWidget *parent = nullptr, Qt::WindowFlags fl = Qt::WindowFlags(), bool embeddedMode = false );
```
instead of a bool here, I'd rather see a "widget mode" enum used

elpaso replied:
Done in https://github.com/qgis/QGIS/pull/4629/commits/8564e8b2b65c7b5aec379a7c5dfb0057076936e7

nyalldawson commented on src/app/qgsdatasourcemanagerdialog.cpp:
```diff
@@ -0,0 +1,170 @@
+  connect( ovl, &QgsOpenVectorLayerDialog::addVectorLayers, this, &QgsDataSourceManagerDialog::vectorLayersAdded );
+
+  // Add data provider dialogs
+  QDialog *dlg;
```
uninitialized

nyalldawson commented on src/app/qgsdatasourcemanagerdialog.cpp:
```diff
@@ -0,0 +1,170 @@
+  // Add data provider dialogs
+  QDialog *dlg;
+
+#ifdef HAVE_POSTGRESQL
```
Could we avoid all this hardcoding and just get the providers direct from the registry?

elpaso replied:
I'll have a second look but I'm afraid that's not currently possible, the different providers dialogs have a (slightly) different interface or need a different setup: see for example AMS that needs the canvas extent, and some (but not all) of the DB that have a progress dialog, it's bit messy and this also answers to the unbound signals reported by @nirvn (I'll try to address that too btw, I tried to use a common function for all DB-like providers, but some of those lack the progress dialog).

nyalldawson commented:
Really nice - I gave it a test drive and it's certainly a big improvement! well done!

guestagain commented:
Hello Folks, I have an additional suggestion to this feature which may or may not be possible or desired by anyone else but here it goes (sorry do not have the skills to assist with development):

The idea is that in the ~"add any layer" dialog that is been developed here, to the right hand side of that dialog there is a check box that you can click on (default = off) that allows you to open the selected layer as a copy of the file you have selected to open, saving it either in the same folder that it is sourced from - by clicking on an additional check box labelled ~"save in same folder with for example '_001' appended to the end of the original filename", or; with a third column with the usual box with the three little dots that allow you to browse to a selected location and set the desired new filename.

If this is not clear I have uploaded two very (very) rough screenshots of how the implementation would look at the following addresses (you may at least have a giggle at the magnificent msPaint design skills on show): 
1) http://imgur.com/7r1FlMq
2) http://imgur.com/6CicFdN

[note: these are annotations on two alternative versions of this ~"add any layer" concept found at these two web locations:
1) https://github.com/qgis/QGIS/pull/4629 (here) and;
2) https://anitagraser.com/2016/07/20/one-add-button-to-rule-them-all/

Aside: a feature like this, as well as an extra check box/directory search that allows you to set the .qml file prior to opening would be great also.

If it is not possible to implement as part of the dialog box been discussed here, perhaps it may be useful for consideration in the usual "add shapefile" dialogs within QGIS.

Good luck and thank you for all your amazing work in any case.

nirvn commented:
I like how this is being implemented, good job.

One thing this new dialog has revealed is the UI inconsistencies between various server provider connection dialogs:
* "Connections" (e.g. PostgreSQL) vs. "Server Connections" (WFS) vs no title (WCS)

nirvn commented:
BTW, console had a couple of warnings popping up:

```
Warning: QObject::connect: No such signal QgsSpatiaLiteSourceSelect::progress( int, int )
Warning: QObject::connect:  (sender name:   'QgsDbSourceSelectBase')
Warning: QObject::connect:  (receiver name: 'QgsDataSourceManagerDialog')
Warning: QObject::connect: No such signal QgsSpatiaLiteSourceSelect::progressMessage( QString )
Warning: QObject::connect:  (sender name:   'QgsDbSourceSelectBase')
Warning: QObject::connect:  (receiver name: 'QgsDataSourceManagerDialog')
Warning: QObject::connect: No such signal QgsMssqlSourceSelect::progress( int, int )
Warning: QObject::connect:  (sender name:   'QgsDbSourceSelectBase')
Warning: QObject::connect:  (receiver name: 'QgsDataSourceManagerDialog')
Warning: QObject::connect: No such signal QgsMssqlSourceSelect::progressMessage( QString )
Warning: QObject::connect:  (sender name:   'QgsDbSourceSelectBase')
Warning: QObject::connect:  (receiver name: 'QgsDataSourceManagerDialog')
Warning: QObject::connect: No such signal QgsDb2SourceSelect::progress( int, int )
Warning: QObject::connect:  (sender name:   'QgsDbSourceSelectBase')
Warning: QObject::connect:  (receiver name: 'QgsDataSourceManagerDialog')
Warning: QObject::connect: No such signal QgsDb2SourceSelect::progressMessage( QString )
Warning: QObject::connect:  (sender name:   'QgsDbSourceSelectBase')
Warning: QObject::connect:  (receiver name: 'QgsDataSourceManagerDialog')
Warning: QObject::connect: No such signal QgsAmsSourceSelect::addRasterLayer( QString const &, QString const &, QString const & )
Warning: QObject::connect:  (sender name:   'QgsSourceSelectBase')
Warning: QObject::connect:  (receiver name: 'QgsDataSourceManagerDialog')
Warning: QMetaObject::connectSlotsByName: No matching signal for on_mDialogButtonBox_helpRequested()
```

nirvn commented:
I'd recommend removing the close button visible in the GIF shot. It saves space, and fits other parts of QGIS (snapping configuration, style manager)

nirvn commented:
Also, I'd put the [+] button as part of the layers panel toolbar. IMHO it's a better suited and more natural placement for it, and it'll allow ppl to hide the main toolbar with individual providers off if not needed.

elpaso commented:
@guestagain thanks for your ideas, this PR is limited in scope to the first phase as described at the end of the linked google doc: embed the browser and the current provider dialogs into a single one.

The second phase will target the browser to make it feature complete (features parity with the provider dialogs).

So, please add your ideas to the google document, or file a feature request on the hub.

elpaso commented:
@nirvn 
> Also, I'd put the [+] button as part of the layers panel toolbar. IMHO it's a better suited and more natural placement for it, and it'll allow ppl to hide the main toolbar with individual providers off if not needed.

I'm sorry but I don't understand what you mean, the "+" button is currently already part of the layer toolbar:

![immagine](https://cloud.githubusercontent.com/assets/142164/26439005/64c0781c-4126-11e7-989d-60774f2ff40e.png)

But, because the actual buttons for the dialogs will be optional (probably they will just stay where they are and the layer toolbar will be hidden by default) I believe we need to put the "+" button in a separate toolbar.

Suggestions welcome!

elpaso commented:
@nirvn 

> I'd recommend removing the close button visible in the GIF shot. It saves space, and fits other parts of QGIS (snapping configuration, style manager)

I see your point but the close button is actually there **for consistency** with the other dialogs (options, layers properties, about, plugin manager etc.), in fact it is required by the base class `QgsOptionsDialogBase` (and I'm going to add an Help button too).

I'm not convinced that from a usability perspective, removing the Close button is a good idea, but in case we take that route, that requirement should be removed from the base class so that we have a consistent GUI.

BTW, your comment brings to the front another important topic: we should write a more comprehensive HIG document.

nirvn commented:
@elpaso , here's what I meant, in image :)
![screenshot from 2017-05-25 14-21-32](https://cloud.githubusercontent.com/assets/1728657/26440303/e4f58a66-4155-11e7-83ca-2f445f81b921.png)

elpaso commented:
@nirvn oh, I see, thanks for the snapshot :)

That makes sense, I'm just worried because that it's a PANEL and could be hidden, but that would be quite an odd setup anyway.

nirvn commented:
@elpaso , in terms of panel hidden, we could have the same argument with the layer management toolbar (i.e., user turning it off). I think in this case, the panel is a better placement 😄

elpaso committed:
[addlayerbutton] Added help button (URL is still a placeholder)

elpaso committed:
[addlayerbutton] Initialise pointer

SrNetoChan commented:
I like @nirvn idea. IMHO, it's a very convenient and also intuitive place to have a button to load data.

About Question 2: I would really like if this could be a dockable panel, hence non-modal dialog. Those with enough screen space could use it docked, while others could use it in floating mode with the exact same functionality it as now.

Also, I was under the impression that this data manager dialog/panel would replace the browser (at least in the long run), no? Otherwise, we are just adding yet another way to load data, which IMHO, may end up being more confusing, and harder to maintain.

DelazJ commented:
@SrNetoChan 
> Also, I was under the impression that this data manager dialog/panel would replace the browser (at least in the long run), no? Otherwise, we are just adding yet another way to load data, which IMHO, may end up being more confusing, and harder to maintain.

This is what I thought too: add to the Browser Panel in a more convenient way the missing features from all the other data loader. And I do not feel it's the case reading some comments and due to the screencast.
Nota: I didn't test the branch yet hence might be missing something

elpaso commented:
@SrNetoChan , @DelazJ 

It should be all in the linked Google doc.

The roadmap is to eventually get rid of the dialogs and keep the browser. But this is not in scope for this PR because the browser is not feature complete.

Phase 1 (current PR) is to hide the individual providers dialogs and toolbar and replace them with this unified dialog with the embedded browser as the default tab.

Phase 2 (future PR, to be assigned) is to add the missing features to the browser and be able to do every single operation from it.

So, all functionality currently missing from the browser will be temporarily available in the dialogs.

> and harder to maintain.

not as long as it uses the browser as an embedded widget (they share the same code).

nirvn commented:
@elpaso , I'm -1 on removing the toolbar. One reason why I think the [+] button should be located in the layer panel is to actually preserve the provider toolbar and be able to have these useful shortcuts at hand. Clicking on the individual provider toolbar icon either opens the individual provider dialog, or opens the unified dialog focused on that specific provider's tab (no strong feeling either way here).

wonder-sk commented:
@elpaso very nice!

Would it be possible to add support for an interface so that plugins can register custom pages for "add layer" dialog? This would be also useful to make the existing widgets more uniform as they would be all using a common interface...

elpaso committed:
[addlayerbutton] Use enum for window mode instead of booleans

elpaso committed:
[addlayerbutton] Added delimited text layers

elpaso commented:
@wonder-sk yes, that would be a nice addition, thanks for raising the point.

Unfortunately there is no common interface for the `Qgs<name your provider>SourceSelect` dialogs, and some even have post configuration. An additional blocker is the complete lack of homogeneous signals and slot naming in the providers (for instance we have `addVectorLayer` but also `addWfsLayer` etc.).

I'm afraid that this refactoring would be out of scope for this PR, but I totally agree with you that we need to eventually address a deeper refactoring of the providers and related dialogs towards a common interface, 

BTW, with this PR I've already started to make the function signatures for the dialogs factory look the same for all providers, at least it is a step in the right direction, see:  https://github.com/qgis/QGIS/pull/4629/commits/8564e8b2b65c7b5aec379a7c5dfb0057076936e7

elpaso committed:
[addlayerbutton] Virtual Layers support added

Some changes in connections are required to keep in sync
the available layers lists in the virtual layer dialog.

(Modeless support is on the way)

SrNetoChan commented:
We have been discussing some aspects of the UI/UX of this new functionality. To allow @elpaso complete this work, we need to make some decisions. Note that the create new layer buttons should not be included in the data source manager, so we will have two buttons to manage. So, here it goes a summary of what we have until now:

**Kinda consensual?:**
1 - Old data management toolbar should not be removed, but only disabled by default;
2 - Menu entries for each provider should not me removed either (to keep shortcuts).

**3 - Where to put the "unified button" and create new layer button?**
   a) In a separate toolbar (to enable turning the rest of the buttons on and off)
![image](https://cloud.githubusercontent.com/assets/3607161/26491575/1ad1aa54-4208-11e7-9601-9787ed07d3ee.png)
![image](https://cloud.githubusercontent.com/assets/3607161/26491534/f5fae088-4207-11e7-9f64-1010cd6bd49c.png)

   b) In the layers panel toolbar.
![image](https://cloud.githubusercontent.com/assets/3607161/26491777/f03b866a-4208-11e7-9c20-65d0068f6e4a.png)
   pros: a convenient place to load data
   cons: Layer panel can be closed; Too many buttons in the layers panel's toolbar;

**4 - If we go with solution 3a, where, by default, should we put the new toolbar?**

**Note:** We are talking about the default, so what matters is the perception new users will have. Others can always customize it the way they like the most.

   a) Left vertical toolbar area (where it is now)

![image](https://cloud.githubusercontent.com/assets/3607161/26491155/4b7e484e-4206-11e7-8b7d-98ea3a38159c.png)

   maybe with some company

![image](https://cloud.githubusercontent.com/assets/3607161/26491984/dae8c22c-4209-11e7-9a69-70c4b02ba012.png)

   b) Somewhere in the top toolbars (so it does not sit alone in the left, wasting space).

![image](https://cloud.githubusercontent.com/assets/3607161/26491095/0685e8d2-4206-11e7-860c-98c18e38c74e.png)

**4 - Clicking the menu items, what should happen?**
    a) Do they open the old/current provider dialogs?
    b) Do they open the new data source manager dialog in the provider's tab?

SrNetoChan commented:
Regarding the menu items..

**5 - where should we put the new Add Layer button?**
 a) In layer > add layer > Open Data source manager.
![image](https://cloud.githubusercontent.com/assets/3607161/26492984/f63d6e74-420e-11e7-8e2d-3421a1ae241f.png)

 b) In layer > Open Data Source Manager.
![image](https://cloud.githubusercontent.com/assets/3607161/26492821/109e5220-420e-11e7-8e43-008234693466.png)
 c) Don't put it on the menus. (may be related to question 2)

**7 - Clicking the menu items, what should happen?**
    a) Open the current provider individual dialogs.
    b) Open the new Data Source Manager dialog in the respective provider's tab.
        (This would replace the need to have a menu entry for opening the data source manager)

elpaso committed:
[addlayerbutton] Raster support added

There is still an issue with the list not being
correclty reset after the raster icon is selected.

DelazJ commented:
Imho, the 3.0 changelog should state that there's now only one place to add layers to QGIS. I mean one dialog *from the User perspective* (whether it's not yet the case in the code base is another issue). So whatever button or menu the user selects, it would open the Data source manager dialog, either on the last tab (using the new entry) or the selected provider tab (using provider's entry).
For the User, the individual modal dialogs should no longer exist. This is a more coherent behavior and might reduce the gap once all this is merged with the Browser panel in the step 2 of the implementation. Also, if we keep showing old dialogs, it will rather looks like we "simply" add another tool instead of reducing the list.
This is the way I conceive these changes.

Trying now to give my opinions for the points @SrNetoChan raised
3a/ yes for a separate toolbar (not add it to current Layers toolbar
b/ I think this placement has a lot of sense. There's a remove layer button so let's add a "Add layer" one. Not sure that the "create layer" button should also go to Layers panel. The fact that the layers panel could be hidden is not really a big argument. Any panel or toolbar  in QGIS can be hidden.
4/ I would say (a) with company but no strong opinion
5/ I'd say (b) - do not mix it other buttons because of toolbar separation. But I'd move it upper in the dialog (2nd place?)
7/ according to my initial comment, (b).

Just to be sure, at the end of the implementation, once all is done, there will only be one button (_to rule them all_ as Anita said), right? We'll remove all the current provider's buttons? --> using the current provider's buttons or menu entries is just a temporary configuration. Am I right?

elpaso committed:
[addlayerbutton] Larry Shaffer's fix to the raster button

Tricky solution, but temporary

elpaso committed:
[addbuttonlayer] Fix layout of vector dialog

This is an attempt to fix the layout of the vector
layer dialog to fit in the unified dialog layout.

The optiomal solution would probably involve
a complete refactoring of the vector layer dialog
to use a stacked widget.

nirvn commented:
@elpaso , @SrNetoChan , OK, lots to catch up to :)

I think we all agree on point 1 and 2, we can look that in.

**Point 3**, I'm honestly a -1 to create a two-buttons main window toolbar. IMHO, that's just messy, I'd rather avoid adding one more toolbar to our 14 preexisting ones :) Like I've stated above, in my view, the cleanest and logical approach is to add it to the layer panel, where we already have a remove layer button. 

That said, you've raised a good point re "too many buttons" in the layer panel's toolbar. It was already an issue before, and it would be could to fix this now that we're on the topic. Two things should be done to improve the situation: a/ move the "expand all", "collapse all" buttons into the [ eye ] dropdown toolbar button (it'll sit well with the show all / hide all layers, etc), b/ merge the confusing presence of two filter buttons into one dropdown toolbar button.

**Point 5**, what about putting the open data source manager to be at the very top of the layer menu? You'd want that the closest to your mouse pointer when expanding that menu.

BTW, If you guys don't have the time / resource to come up with an icon for this feature, I can commit one once this PR is merged. Sharing the burden 😉 

**Point 7**, I'd pick option b, namely the menu actions open the data source manager with a focus on the clicked provider tab. I do not see any point in keeping individual dialogs around when we have something that's superior and doesn't get in the way. That'll streamline the UX.

shenriod commented:
2 cents: I am much more in favor of 3a) than 3b). Beginners could very easily close the layers panel by mistake and would immediately feel lost "how can I add a new layer?" I see it much less likely that they disable the toolbar by mistake.

However I don't think that both are mutually exclusive. Any strong reason why we could not have 3a) AND 3b) ? I don't see this as a  bad redundancy

I would also favor 4b) over 4a)

For 5, I agree with @nirvn : "what about putting the open data source manager to be at the very top of the layer menu? You'd want that the closest to your mouse pointer when expanding that menu."

nirvn commented:
(I should have noted that the new layer button shouldn't be in the layer panel, that's a bit much :) )

elpaso committed:
[addlayerbutton] Removed help button

Almost all provider dialogs have its own button
resulting in duplicated help buttons in the
unified dialog.

elpaso committed:
[addlayerbutton] Added option for modeless layer dialog (default false)

elpaso commented:
Whatever we will agree upon, I'll try to put the whole menu thing in a single commit, so that we can revert easily if we decide to change anything.

Coming to the points.

I agree with @nirvn for point 5 and 7 but I'm still convinced that the add layer icon does not belong to the layer panels menus (and I'd go for **4/a** a 2-button separate toolbar, that will replace the current one (which in turn will be disabled by default) and this is why:
- the layer panel toolbar is focused on canvas actions, it's a legend toolbar: it deals with organizing, removing switching on/off layer in the canvas 
-  it is true that the new data source manager dialog can also add layers to the legend and to the canvas but it also has data source management capabilities (for example, create a new connection etc.)
- icons in the main vertical toolbar are more visible and larger than those in the already crowded layers panel menu
- toolbars can be easily customized (moved re-arranged) I'm not sure if we can do the same for the layer's panel toolbar

About adding the same icon in two places proposed by @shenriod , it is not generally a good idea to add the same functionality in more than one place but I understand that there could be exceptions to the rules by the book.

... and thanks @nirvn  for your offer with the icons I definitely need help with those icons, I'll leave the green plus icon for now but that's just a placeholder, feel free to propose a real icon.

elpaso committed:
[addlayerbutton] Add method to open a given page in the layer dlg

This is preliminary to change all dialogs slots to
open the unified dialog with the righ page open.

elpaso committed:
[addlayerbutton] Fix order of signals when open page called

The problem was with raster "tab" (that is not a real page
but a forwarded signal to the app), when openPage
was called before the show() call.

Moving the activation inside the unique currentPage call,
the implementation is much simpler (got rid of the lambda
with the second currentRowChanged connection).

elpaso committed:
[addlayerbutton] Removed all the individual source select slots

In favour of the unified dialog.

elpaso committed:
Removed old comment

elpaso committed:
[addlayerbutton] New GUI for the delimited text dlg by Luigi Pirelli

luipir commented:
Just a note about "delimited text" gui. It has been reorganized to be cleaner and to fit better inside SingleAddButton GUI. The guideline has been to avoid remove any existing GUI element  that is referred with objectNames.
The GUI can be simplified more (e.g. CSV can be merged with custom + ","). Btw it is just a fist step because GUI reorganization wasn't in the scope of this phase.

elpaso committed:
[addlayerbutton] Data Source Manager toolbar and menu

This implements the proposal from @SrNeto:

- new toolbar with the new dialog button and
  all the "new" data soource buttons
- old layer toolbar hidden by default
- new dialog open from first item in layer menu

elpaso commented:
Implementation is finished, give it a test ride!

I've implemented the new toolbar and menu item in the single last commit, that can be easily reverted if we agree upon a different way to acess the dialogs and menus.

nirvn commented:
If we end up settling on a dedicated toolbar for this, we should move the db manager toolbar icon away from its mono-button "database toolbar" into this newly created toolbar.

nirvn commented:
@elpaso , testing your latest commits, we're getting really close to a final product here, nice.

There's a problem with the delimited text UI, whereas it should be embedded into a scrollable area, to prevent the dialog from taking too much vertical space. Currently, the *minimum* height of the dialog is way to tall:
![screenshot from 2017-05-30 08-48-40](https://cloud.githubusercontent.com/assets/1728657/26565398/14aaadf0-4515-11e7-8106-74fdee5725e3.png)

Also, to continue a discussion we had above, I'd still be pushing for the close button to be removed. It really doesn't look at good at the moment, with the [ help ] and [ add ] buttons row which we expect to be at the bottom showing up above close:
![screenshot from 2017-05-30 08-51-30](https://cloud.githubusercontent.com/assets/1728657/26565431/61fcab44-4515-11e7-84e4-183264ea8be2.png)

@luipir , nice job on improving the delimited text UI.

nirvn commented:
Oh, lastly, we can't have the "new [insert layer type] layer" buttons in both the layer management toolbar and this new toolbar (if we settle on that ;-P), we'll have to decide where they go.

nyalldawson commented:
My 2c, based on tests of the most recent commit:

- I think the browser tab needs an explicit "Add layer" button. It's not obvious for users that they need to right-click layers and then add them to the project, and the small toolbar button (which incidentally looks more like "new layer") is also not obvious. I'd suggest moving the "Add" button down to the button bar next to the "Close" button, and hooking it up so it triggers the add functionality from each separate tab (this also would fix @nirvn's concerns about the double button bar)
- I don't think clicking the raster tab should immediately pop up the file browser. This differs in behavior considerably from the other tabs. It's also a bit odd - because selecting a file then throws the dialog across to the Postgres tab. We probably need a page like the ogr page, containing the browse button.
- @nirvn I think we need versions of the sidebar icons without the green plus for this dialog. Do you agree?Also that db2 icon looks horrendous here with its non-transparent background
- Is there any logic to the order of providers listed?

nyalldawson commented:
I should preface that whole comment by saying I love the direction it's going and think in general it's working really well! :100:

nirvn commented:
@elpaso , @nyalldawson , I'm also not convinced by the immediate file picker dialog pop up for the raster provider, +1 to implement the simplest of UI (basically, a source group box with a line edit and a [ browser ] button).

elpaso commented:
@nirvn 

> Oh, lastly, we can't have the "new [insert layer type] layer" buttons in both the layer management toolbar and this new toolbar (if we settle on that ;-P), we'll have to decide where they go:

I think that what you are asking for is already how I implemented that, this is how it looks like now:

![unibutton4](https://cloud.githubusercontent.com/assets/142164/26569169/4d59a784-4509-11e7-829d-442423398dfb.png)

elpaso commented:
@nyalldawson 
> My 2c, based on tests of the most recent commit:

> I'd suggest moving the "Add" button down to the button bar next to the "Close" button, and hooking it up so it triggers the add functionality from each separate tab (this also would fix @nirvn's concerns about the double button bar)

Sorry but this is out of scope for this phase because that would need a complete refactoring of most dialogs: the individual dialogs have completely different toolbars. I've already done some homogenization, changing the button label to _Add_ and changing the `accept()` to not close the dialog but I've no room for a complete refactoring in this phase (I'm pretty sure I'll be able to continue working on this features in the future though).

Or maybe I misunderstood and you are proposing to move only the _Add_ button down into the main dialog button bar? In that case the double button row would not go away.

But I totally I agree that the double row sucks. What can be more easily implemented is probably to add a Close button to all embedded dialogs and forward it to the parent's `accept()`.

>    I don't think clicking the raster tab should immediately pop up the file browser. This differs in behavior considerably from the other tabs. It's also a bit odd - because selecting a file then throws the dialog across to the Postgres tab. We probably need a page like the ogr page, containing the browse button.

Yeah, that can be done adding a push button inside the raster stacked widget page. I did not do that because it would require an extra mouse click and you really get no extra functionality.

(I don't get what you mean with "...throws the dialog across the postgres tab": the intended behavior is that the active tab will be reset to whatever it was before the user clicked on the raster icon)

>    Is there any logic to the order of providers listed?

My intention was to replicate the existing order. I've no idea about what was the criteria for the original sorting.

nirvn commented:
> I think that what you are asking for is already how I implemented that, this is how it looks like now:

@elpaso , sorry, my comment might not have been clear enough. What I was referring to is the presence of "new memory/shapefile/etc. layer" actions in both the new "Data Source Manager Toolbar" you've added in this PR as well as the pre-existing "Manage Layers Toolbar". See the screenshot below:
![screenshot from 2017-05-30 08-58-21](https://cloud.githubusercontent.com/assets/1728657/26570978/878c4b6c-453d-11e7-9b0c-af9634bc8c76.png)

NathanW2 commented:
I know some people prefer it but I'm really not keen on having yet another
toolbar, I would even say this stuff should live on the project toolbar.
Having another toolbar will just introduce confusion more than what is
currently there.   We can have the add button on the layers toolbar and
just hide all the other behind a single dropdown for now?

On Tue, May 30, 2017 at 4:40 PM, Mathieu Pellerin <notifications@github.com>
wrote:

> I think that what you are asking for is already how I implemented that,
> this is how it looks like now:
>
> @elpaso <https://github.com/elpaso> , sorry, my comment might not have
> been clear enough. What I was referring to is the presence of "new
> memory/shapefile/etc. layer" actions in both the new "Data Source Manager
> Toolbar" you've added in this PR as well as the pre-existing "Manage Layers
> Toolbar". See the screenshot below:
> [image: screenshot from 2017-05-30 08-58-21]
> <https://cloud.githubusercontent.com/assets/1728657/26570978/878c4b6c-453d-11e7-9b0c-af9634bc8c76.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/4629#issuecomment-304788207>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3M1aiV-c-zGQPVoQjwDg9mvH1W4cks5r-7n1gaJpZM4Nk-e6>
> .
>

nirvn commented:
@elpaso , on the two-row situation, if you are running out of allocated time for this, I'd simply get rid of the [ close ] button (per my initial suggestion). We never had [ close ] buttons on the individual provider dialogs before, and while other windows using the base class you're employing here do have such buttons ( layer property, options, etc. ), i really don't think its absence hurt the UX here.

elpaso commented:
@nirvn

> "new memory/shapefile/etc. layer" 

that's on purpose: the old toolbar is disabled by default and replaced by the new one. The old toolbar is there for (possibly nostalgic experienced users) who want it back and save some clicks.

Also, on removing the two rows issue, you won :) that's the fastest way to fix it. BTW it's not correct that we never had a `Close` button in the provider dialogs: a `Close` or a `Cancel` button is present in all of them.

@NathanW2 I don't think we'll find two persons that think the same way here ;) and I have no strong opinion but and I don't want that to be a blocker. I guess we should merge even if we do not reach full consent and discuss further improvements in later PRs after some testing of the GUI, I'll keep the whole menu/toolbar stuff in a single commit, easy to revert.

The points in my reasoning are:
- provide a sensible default toolbars configuration (the new toolbar)
- offer an easy way to get the old configuration back (the old toolbar) because it can save some clicks and it's important for experienced users ( @NathanW2  if we group the old icons in a dropdown we would not spare any click )

nirvn commented:
> Also, on removing the two rows issue, you won :) that's the fastest way to fix it.

🎉 :)

elpaso committed:
[addlayerbutton] Removed button bar from main dialog

No Close button anymore ... prepare your wrist for
a precise hit on the [X] or hit ESC to close the
dialog.

SrNetoChan commented:
I like what we have now. I think our main goal was achieved; we now have a cleaner interface that intuitively points the inexperient user to loading data. But we kept old behaviors so that users less inclined for big changes or aiming for the minimal number of clicks can still have their old toolbar back easily.

> If we end up settling on a dedicated toolbar for this, we should move the DB Manager toolbar icon away from its mono-button "database toolbar" into this newly created toolbar.

@nirvn I'm sure I would like to have the database manager button in here for two reasons.
1 - I think we should keep this toolbar as clean as possible by only having buttons that are useful for the majority of the users.
2- The database toolbar that is a mono-button bar by default, should still exist to receive the buttons for any "Database" plugins.

I would be +1 for adding a duplicated button to the layer panel, as long as we implement @nirvn suggestions to clean that panel toolbar (which might make sense anyway):

> Two things should be done to improve the situation: a/ move the "expand all", "collapse all" buttons into the [ eye ] dropdown toolbar button (it'll sit well with the show all / hide all layers, etc), b/ merge the confusing presence of two filter buttons into one dropdown toolbar button.

@nyalldawson on the Raster dialog. I agree that it feels strange, but adding a line edit and the browser button will just add extra clicking. I think that, in the future, we will want to remove both Vector and Raster tabs from this dialog and simply use the browser.

There are one or two issues that @elpaso is taking care of (crash on loading projects with the browser tab and the extra scrolling in the delimited text dialog). After that, I think we could try to merge this work so that more users can give it a spin from the nightly builds repo. We will surely receive more feedback and we can see if some of it could be done still in phase 1.

elpaso committed:
[addlayerbutton] Prevent a crash when the layer list is rebuilt after a change in the layer tree

elpaso committed:
[addlayerbutton] Fix height problems within the delimited text dialog

Also moved the status information under the button bas for
consistency with other dialogs.

Cracert commented:
Sorry to join this issue so late.

Very first version of 'add layer' icon was this one:
![layer-add](https://cloud.githubusercontent.com/assets/630667/26677527/ad33ad3e-46cc-11e7-8768-7a6afa2c961d.png)
https://github.com/Cracert/GIS-icons/blob/master/24x24/layer-add.png

It supposed to be very generic. The SVG version is also available and I will upload it soon. Let me know is some modifications could improve usability of this icon.

elpaso commented:
@Cracert thanks, if you can upload the svg I'll replace the actual icon with the new one.

Cracert commented:
Thank you @elpaso for this great work! 

Two SVG versions are in this commit, but for me, original one looks better.
https://github.com/Cracert/GIS-icons/commit/d5bff3e751a68ae073cee2ea25e00ec60d30c94a

* layer-add.svg (original)
* layer-add-v2.svg (just to test bigger layer part)

elpaso committed:
[addlayerbutton] Replace icon for the datasourcemanager

elpaso committed:
[addlayerbutton] Connect and forward all signals from browser to app

shenriod commented:
Thanks all who contributed to this new feature!

A few comments and requests for opinions after a few tests:

**Behavior of the Raster provider**
I agree with @nyalldawson and @nirvn  that the raster provider should not open directly the browser. A very simple UI (only with the browse button?) would probably be more intuitive for the users

**DB2 Icon with non-transparent background**
I checked the svg here https://github.com/qgis/QGIS/blob/master/images/themes/default/mIconDb2.svg and here https://github.com/qgis/QGIS/blob/master/images/themes/default/mActionAddDb2Layer.svg and they seem to have a transparent background. Thus not sure why it's not being rendered correctly?

**Order of the icons**
I would suggest a re-ordering of the icons (and slight renaming), so that all file-based types are together, all db-based together, etc...

What about:
- Browser
- Vector (not _Vector files_)
- Delimited Text
_[separation line]_
- PostgreSQL
- Spatialite
- MSSQL
- DB2
_[separation line]_
- WMS
- WFS
- WCS
- ArcGIS Map Server
- ArcGIS Feature Server
_[separation line]_
- Virtual layer (not _Virtual_)

Any opinion / objection on this?

Cheers!

nirvn commented:
@elpaso , after giving this nice UX upgrade a test run for a week now, I really think the raster provider needs to have a simple [ browse ] UI ala OGR vector provider. It's a bit irritating to re-open the add layer button and not have it stick to raster when it's the last used provider.

Plus, pasting an URL for a raster (i.e. /home/gisuser/rasters/my.tif) in a text box can be quicker than using the file picker dialog 😄

elpaso commented:
@nirvn no objections from my side.

DelazJ commented:
From my side, something that puzzles me is that one of the main target of this restructuring (or at the end of the whole process) is to give more visibility to the QGIS Browser. And while experimenting the Data source manager dialog, I almost never use the embedded browser. None of the existing icons opens that tab. In other words, i'm afraid the new dialog is hiding the qgis browser...
How about  keep opening the corresponding tab when one of the usual tools is used and opening the browser tab when someone hits the add layer button itself (instead of the latest opened tab) ? Not sure though it'd be a nice UX. But it could also be a temporary state (until the total revamp).

luipir commented:
@DelazJ Hide Browser? You proposal is to go back to Browser everytime? Can be a simply a configurable behavior (previous tab or tab <X> as default and by default Browser tab) ?
